### ISSUE TITLE
dense sparse dense training

### ISSUE THREAD

**Issue by marsupialtail** [NervanaSystems/distiller#315]
Hi, I am wondering if distiller has a way to specify a schedule that **decreases** the sparsity level of a neural network, initializing the new weights at zero. This is to implement the dense sparse dense training technique of Han et al. 

If this is not possible, can you give me some pointers on how to add a pruner scheduler class? Can I just copy a pruner class I currently find? Do I have to specify the name of the new class somewhere else in the source code? Thank you!

**Comment by guyjacob** [comment]
Apologies for the delayed reply.

I don't believe we have "native" support for dense-sparse-dense using a single schedule (take a look at #76).
However, a while ago we added the option to load just the parameters of the model from a checkpoint, and nothing else (specifically, it will not the pruning masks). So you can execute a pruning session, save a checkpoint, and then do a separate run where you load the model parameters checkpoint and continue to train without any pruning.
If you're running `compress_classifier.py`, you can use the `--exp-load-weights-from <path-to-checkpoint>` command line option. If you're running your own script, you can call `distiller.apputils.load_lean_checkpoint` instead of `distiller.apputils.load_checkpoint`.
If needed you can couple that with a dedicated scheduler to control LR scheduling.

Hopefully this gets you what you need. I don't think you need to implement a new pruner for that.

Just in case I missed something, I'm assigning @nzmora to this. Please note he won't be able to respond this week.

**Comment by nzmora** [comment]
Hi @marsupialtail,

@guyjacob's suggestion for "implementing" DSD is sound.  I want to point out some things:
1. You mention that you're looking for a schedule that **decreases** sparsity.  You don't need this for DSD, which does (you can repeat this several times):
   - Train the DNN
   - Prune the weights below some value (we prune by maintaining a mask on the weights)
   - Fine-tune the DNN while keeping the mask
   - Discard the mask and train at 1/10 the learning rate
2. I think that in DSD they do one-shot pruning (which Distiller supports), but I don't remember if they threshold the weights of all layers at once (which we don't do).
3. It's easy to discard the weights programatically (instead of loading using ```--exp-load-weights-from <path-to-checkpoint``` as @guyjacob suggested): The pruning masks are stored in an instance of ```CompressionScheduler``` which is loaded from the checkpoint.   But you should feel free to change the maskers in the ```CompressionScheduler``` in any way you see fit.  Specifically you can set them to ```None```.   See: https://github.com/NervanaSystems/distiller/blob/5c2e9f54b052cc8eefaf109ec8d52537d41b1e34/distiller/scheduler.py#L81-L94).
4. There are several examples of programmatically using the Distiller APIs outside of the "standard" pruning schedule loop.  E.g. in [SA](https://github.com/NervanaSystems/distiller/blob/43548debc7e8aea743889278d0f1076eeaadc67e/distiller/sensitivity.py) and [Greedy pruning](https://github.com/NervanaSystems/distiller/blob/43548debc7e8aea743889278d0f1076eeaadc67e/distiller/pruning/greedy_filter_pruning.py).
I hope this extra info helped.
Neta
